File: pandora_vision/pandora_vision_hole_detector/test/functional/hole_detector_test.py
#!/usr/bin/env python
# Software License Agreement
__version__ = "0.0.1"
__status__ = "Production"
__license__ = "BSD"
__copyright__ = "Copyright (c) 2014, P.A.N.D.O.R.A. Team. All rights reserved."
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of P.A.N.D.O.R.A. Team nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
__author__ = "Tsirigotis Christos"
__maintainer__ = "Tsirigotis Christos"
__email__ = "tsirif@gmail.com"

PKG = 'pandora_vision_hole_detector'
NAME = 'hole_detector_test'

import sys
import math

import unittest

import roslib; roslib.load_manifest(PKG)
import rostest
import rospy

import cv2

from pandora_testing_tools.testing_interface import test_base
from pandora_vision_msgs.msg import HolesDirectionsVectorMsg
from pandora_vision_msgs.msg import CandidateHolesVectorMsg
from pandora_vision_msgs.msg import EnhancedHolesVectorMsg

class HoleDetectorTest(test_base.TestBase):

    # tests the function of the hole detector package
    def test_hole_detector(self):

        self.assertTrue(self.replied)

        width = 640
        height = 480

        holesDirectionsMsgs = self.messageList[0]

        # how many alert messages have been published
        self.assertEqual(len(holesDirectionsMsgs), 1)

        # how many alerts have been sent with the first message
        self.assertEqual(len(holesDirectionsMsgs[0].holesDirections), 1)
        hole_0 = holesDirectionsMsgs[0].holesDirections[0]

        # the coordinates of the first valid hole's keypoint
        x = 352.86
        y = 238.17

        # the coordinates of the first valid hole's keypoint relative to the
        # center of the image
        expectedX = x - width / 2
        expectedY = height / 2 - y

        # the hole's expected yaw
        expectedYaw = math.atan((2 * float(expectedX) / width) * math.tan(58 * math.pi / 360))

        # the hole's expected pitch
        expectedPitch = math.atan((2 * float(expectedY) / height) * math.tan(45 * math.pi / 360))

        self.assertAlmostEqual(hole_0.yaw, expectedYaw, 2)
        self.assertAlmostEqual(hole_0.pitch, expectedPitch, 2)
        self.assertGreater(hole_0.probability, 0.5)



    # tests the function of the depth node
    def test_depth_node(self):

        self.assertTrue(self.replied)

        candidateHolesMsgs = self.messageList[1]

        # how many alert messages have been published
        self.assertEqual(len(candidateHolesMsgs), 2)

        # how many alerts have been sent with the first message
        self.assertEqual(len(candidateHolesMsgs[0].candidateHoles), 2)

        # make assertions about the first candidate hole
        hole_0 = candidateHolesMsgs[0].candidateHoles[0]

        # the coordinates of the first valid hole's data
        # are halved because of the wavelet analysis used

        # The first hole's keypoint
        x = hole_0.keypointX * 2
        self.assertLess(356, x);
        self.assertGreater(358, x);

        y = hole_0.keypointY * 2
        self.assertLess(239, y);
        self.assertGreater(241, y);

        # There should be equal number of elements in both vertices vectors
        self.assertEqual(len(hole_0.verticesX), len(hole_0.verticesY))

        # The first hole's bounding box is itself bounded
        # in a bounding box
        for i in range(0, len(hole_0.verticesX)):
            self.assertLess(260, 2 * hole_0.verticesX[i])
            self.assertGreater(450, 2 * hole_0.verticesX[i])

        for i in range(0, len(hole_0.verticesY)):
            self.assertLess(195, 2 * hole_0.verticesY[i])
            self.assertGreater(285, 2 * hole_0.verticesY[i])

        # There should be equal number of elements in both outline vectors
        self.assertEqual(len(hole_0.outlineX), len(hole_0.outlineY))

        # The minimum x coordinate of the bounding box's points
        minX = min(hole_0.verticesX)

        # The minimum x coordinate of the bounding box's points
        minY = min(hole_0.verticesY)

        # The maximum x coordinate of the bounding box's points
        maxX = max(hole_0.verticesX)

        # The maximum x coordinate of the bounding box's points
        maxY = max(hole_0.verticesY)

        # The first hole's outline is bounded by its bounding box
        for i in range(0, len(hole_0.outlineX)):
            self.assertGreaterEqual(hole_0.outlineX[i], minX)
            self.assertLessEqual(hole_0.outlineX[i], maxX)


        # make assertions about the second candidate hole
        hole_1 = candidateHolesMsgs[0].candidateHoles[1]

        # the coordinates of the second valid hole's data
        # are halved because of the wavelet analysis used

        # The second hole's keypoint
        x = hole_1.keypointX * 2
        self.assertLess(330, x);
        self.assertGreater(340, x);

        y = hole_1.keypointY * 2
        self.assertLess(100, y);
        self.assertGreater(110, y);

        # There should be equal number of elements in both vertices vectors
        self.assertEqual(len(hole_1.verticesX), len(hole_1.verticesY))

        # The second hole's bounding box is itself bounded in a bounding box
        for i in range(0, len(hole_1.verticesX)):
            self.assertLess(260, 2 * hole_1.verticesX[i])
            self.assertGreater(410, 2 * hole_1.verticesX[i])

        for i in range(0, len(hole_1.verticesY)):
            self.assertLess(70, 2 * hole_1.verticesY[i])
            self.assertGreater(140, 2 * hole_1.verticesY[i])

        # There should be equal number of elements in both outline vectors
        self.assertEqual(len(hole_1.outlineX), len(hole_1.outlineY))

        # The minimum x coordinate of the bounding box's points
        minX = min(hole_1.verticesX)

        # The minimum x coordinate of the bounding box's points
        minY = min(hole_1.verticesY)

        # The maximum x coordinate of the bounding box's points
        maxX = max(hole_1.verticesX)

        # The maximum x coordinate of the bounding box's points
        maxY = max(hole_1.verticesY)

        # The first hole's outline is bounded by its bounding box
        for i in range(0, len(hole_1.outlineX)):
            self.assertGreaterEqual(hole_1.outlineX[i], minX)
            self.assertLessEqual(hole_1.outlineX[i], maxX)


    # tests the function of the rgb node
    def test_rgb_node(self):

        self.assertTrue(self.replied)

        candidateHolesMsgs = self.messageList[2]

        # how many alert messages have been published
        self.assertEqual(len(candidateHolesMsgs), 2)

        # how many alerts have been sent with the first message
        self.assertEqual(len(candidateHolesMsgs[1].candidateHoles), 2)

        # make assertions about the first candidate hole
        hole_0 = candidateHolesMsgs[1].candidateHoles[0]

        # the coordinates of the first valid hole's data
        # are halved because of the wavelet analysis used

        # The first hole's keypoint
        x = hole_0.keypointX * 2
        self.assertLess(345, x);
        self.assertGreater(348, x);

        y = hole_0.keypointY * 2
        self.assertLess(236, y);
        self.assertGreater(238, y);

        # There should be equal number of elements in both vertices vectors
        self.assertEqual(len(hole_0.verticesX), len(hole_0.verticesY))

        # The first hole's bounding box is itself bounded
        # in a bounding box
        for i in range(0, len(hole_0.verticesX)):
            self.assertLess(255, 2 * hole_0.verticesX[i])
            self.assertGreater(440, 2 * hole_0.verticesX[i])

        for i in range(0, len(hole_0.verticesY)):
            self.assertLess(194, 2 * hole_0.verticesY[i])
            self.assertGreater(280, 2 * hole_0.verticesY[i])

        # There should be equal number of elements in both outline vectors
        self.assertEqual(len(hole_0.outlineX), len(hole_0.outlineY))

        # The minimum x coordinate of the bounding box's points
        minX = min(hole_0.verticesX)

        # The minimum x coordinate of the bounding box's points
        minY = min(hole_0.verticesY)

        # The maximum x coordinate of the bounding box's points
        maxX = max(hole_0.verticesX)

        # The maximum x coordinate of the bounding box's points
        maxY = max(hole_0.verticesY)

        # The first hole's outline is bounded by its bounding box
        for i in range(0, len(hole_0.outlineX)):
            self.assertGreaterEqual(hole_0.outlineX[i], minX)
            self.assertLessEqual(hole_0.outlineX[i], maxX)


        # make assertions about the second candidate hole
        hole_1 = candidateHolesMsgs[1].candidateHoles[1]

        # the coordinates of the second valid hole's data
        # are halved because of the wavelet analysis used

        # The second hole's keypoint
        x = hole_1.keypointX * 2
        self.assertLess(320, x);
        self.assertGreater(330, x);

        y = hole_1.keypointY * 2
        self.assertLess(100, y);
        self.assertGreater(110, y);

        # There should be equal number of elements in both vertices vectors
        self.assertEqual(len(hole_1.verticesX), len(hole_1.verticesY))

        # The second hole's bounding box is itself bounded in a bounding box
        for i in range(0, len(hole_1.verticesX)):
            self.assertLess(257, 2 * hole_1.verticesX[i])
            self.assertGreater(400, 2 * hole_1.verticesX[i])

        for i in range(0, len(hole_1.verticesY)):
            self.assertLess(70, 2 * hole_1.verticesY[i])
            self.assertGreater(145, 2 * hole_1.verticesY[i])

        # There should be equal number of elements in both outline vectors
        self.assertEqual(len(hole_1.outlineX), len(hole_1.outlineY))

        # The minimum x coordinate of the bounding box's points
        minX = min(hole_1.verticesX)

        # The minimum x coordinate of the bounding box's points
        minY = min(hole_1.verticesY)

        # The maximum x coordinate of the bounding box's points
        maxX = max(hole_1.verticesX)

        # The maximum x coordinate of the bounding box's points
        maxY = max(hole_1.verticesY)

        # The first hole's outline is bounded by its bounding box
        for i in range(0, len(hole_1.outlineX)):
            self.assertGreaterEqual(hole_1.outlineX[i], minX)
            self.assertLessEqual(hole_1.outlineX[i], maxX)



    def test_enhanced_holes(self):

        self.assertTrue(self.replied)

        enhancedHolesMsgs = self.messageList[3]

        # how many alert messages have been published
        self.assertEqual(len(enhancedHolesMsgs), 1)

        # how many alerts have been sent with the first message
        self.assertEqual(len(enhancedHolesMsgs[0].enhancedHoles), 1)

        # depth analysis is possible
        self.assertTrue(enhancedHolesMsgs[0].isDepth)

        # the first hole
        hole_0 = enhancedHolesMsgs[0].enhancedHoles[0]

        # The first hole's keypoint
        x = hole_0.keypointX
        self.assertLess(350, x);
        self.assertGreater(360, x);

        y = hole_0.keypointY
        self.assertLess(230, y);
        self.assertGreater(240, y);

        # There should be equal number of elements in both vertices vectors
        self.assertEqual(len(hole_0.verticesX), len(hole_0.verticesY))

        # The first hole's bounding box is itself bounded
        # in a bounding box
        for i in range(0, len(hole_0.verticesX)):
            self.assertLess(255, hole_0.verticesX[i])
            self.assertGreater(450, hole_0.verticesX[i])

        for i in range(0, len(hole_0.verticesY)):
            self.assertLess(190, hole_0.verticesY[i])
            self.assertGreater(290, hole_0.verticesY[i])



if __name__ == '__main__':

    subscriber_topics = [
        ("/vision/holes_direction_alert", "pandora_vision_msgs", "HolesDirectionsVectorMsg"),
        ("/pandora_vision/hole_detector/synchronized/depth/candidate_holes", "pandora_vision_msgs", "CandidateHolesVectorMsg"),
        ("/pandora_vision/hole_detector/synchronized/rgb/candidate_holes", "pandora_vision_msgs", "CandidateHolesVectorMsg"),
        ("/vision/enhanced_holes", "pandora_vision_msgs", "EnhancedHolesVectorMsg")]

    rospy.sleep(15)
    rospy.init_node(NAME, anonymous=True, log_level=rospy.DEBUG)
    HoleDetectorTest.connect(subscriber_topics, [])
    HoleDetectorTest.playFromBag(block = True)
    rospy.sleep(2)
    rostest.rosrun(PKG, NAME, HoleDetectorTest, sys.argv)
    HoleDetectorTest.disconnect()

File: pandora_vision/pandora_vision_hole/test/unit/depth_node/hole_detector_test.cpp
/*********************************************************************
 *
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2014, P.A.N.D.O.R.A. Team.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of the P.A.N.D.O.R.A. Team nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 * Author: Alexandros Philotheou
 *********************************************************************/

#include "depth_node/hole_detector.h"
#include "gtest/gtest.h"

namespace pandora_vision
{
  /**
    @class HoleDetectorTest
    @brief Tests the integrity of methods of class HoleDetector
   **/
  class HoleDetectorTest : public ::testing::Test
  {
    protected:
      HoleDetectorTest() {}

      /**
        @brief Constructs a rectangle of width @param x and height of @param y.
        @param[in] upperLeft [const cv::Point2f&] The upper left vertex of the
        rectangle to be created
        @param[in] x [const int&] The rectangle's width
        @param[in] y [const int&] The rectangle's height
        @param[in] depthIn [const float&] The depth value for all points inside
        the rectangle
        @param[out] image [cv::Mat*] The image on which the rectangle will be
        imprinted on
        return void
       **/
      void generateDepthRectangle(
        const cv::Point2f& upperLeft,
        const int& x,
        const int& y,
        const float& depthIn,
        cv::Mat* image );

      // Sets up one image: squares_,
      // which features three squares of size 100.
      // The first one (order matters here) has its upper left vertex at
      // (100, 100),
      // the second one has its upper right vertex at (WIDTH - 3, 3)
      // (so that the blob it represents can barely be identified)
      // and the the third one has its lower right vertex at
      // (WIDTH - 1, HEIGHT - 1)
      virtual void SetUp()
      {
        WIDTH = 640;
        HEIGHT = 480;

        // The image upon which the squares will be inprinted
        squares_ = cv::Mat::zeros(HEIGHT, WIDTH, CV_32FC1);

        // Construct the squares_ image

        // Set the depth for each point of the squares_ image to 1.0
        for (int rows = 0; rows < squares_.rows; rows++)
        {
          for (int cols = 0; cols < squares_.cols; cols++)
          {
            squares_.at<float>(rows, cols) = 1.0;
          }
        }

        // Construct the lower right square
        cv::Mat lowerRightSquare = cv::Mat::zeros(HEIGHT, WIDTH, CV_32FC1);

        HoleDetectorTest::generateDepthRectangle
          (cv::Point2f(WIDTH - 100, HEIGHT - 100),
            100,
            100,
            1.2,
            &lowerRightSquare);

        // Construct the upper right image
        cv::Mat upperRightSquare = cv::Mat::zeros(HEIGHT, WIDTH, CV_32FC1);

        HoleDetectorTest::generateDepthRectangle
          (cv::Point2f(WIDTH - 103, 3),
            100,
            100,
            2.2,
            &upperRightSquare);

        // Construct the upper left square
        cv::Mat upperLeftSquare = cv::Mat::zeros(HEIGHT, WIDTH, CV_32FC1);

        // Construct the square_ image
        HoleDetectorTest::generateDepthRectangle
          (cv::Point2f(100, 100),
            100,
            100,
            1.8,
            &upperLeftSquare);

        // Synthesize the final squares_ image
        squares_ += lowerRightSquare + upperRightSquare + upperLeftSquare;

      }
      // The images' width and height
      int WIDTH;
      int HEIGHT;

      // The image that will be used to locate blobs in
      cv::Mat squares_;
  };



  /**
    @brief Constructs a rectangle of width @param x and height of @param y.
    @param[in] upperLeft [const cv::Point2f&] The upper left vertex of the
    rectangle to be created
    @param[in] x [const int&] The rectangle's width
    @param[in] y [const int&] The rectangle's height
    @param[in] depthIn [const float&] The depth value for all points inside
    the rectangle
    @param[out] image [cv::Mat*] The image on which the rectangle will be
    imprinted on
    return void
   **/
  void HoleDetectorTest::generateDepthRectangle(
    const cv::Point2f& upperLeft,
    const int& x,
    const int& y,
    const float& depthIn,
    cv::Mat* image)
  {
    // Fill the inside of the desired rectangle with the @param depthIn provided
    for (int rows = upperLeft.y; rows < upperLeft.y + y; rows++)
    {
      for (int cols = upperLeft.x; cols < upperLeft.x + x; cols++)
      {
        image->at<float>(rows, cols) = depthIn;
      }
    }
  }



  //! Tests HoleDetector::findHoles
  TEST_F(HoleDetectorTest, findHolesTest)
  {
    // Run HoleDetector:findHoles
    HolesConveyor conveyor = HoleDetector::findHoles(squares_);

    // The number of keypoints found
    int size = conveyor.size();

    // There should be two keypoints: the one of the upper left square
    // and the one of the upper right square. The lower right square is
    // adjacent to the edges of the image and will be clipped by the
    // edge contamination method
    ASSERT_EQ(2, size);

    // For every keypoint found, make assertions and expectations
    for (int k = 0; k < size; k++)
    {
      // The location of the keypoint should be near the center of the square
      // in which it lies
      EXPECT_NEAR(conveyor.holes[k].keypoint.pt.x,
        conveyor.holes[k].rectangle[0].x + 50, 1);

      // The hole should have exactly four vertices
      EXPECT_EQ (4, conveyor.holes[k].rectangle.size());

      // There should be 400 outline points
      EXPECT_EQ (400, conveyor.holes[k].outline.size());
    }
  }

}  // namespace pandora_vision
